System.Device.Gpio/GpioController.cs

```diff
@@ -0,0 +1,268 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections;
+using System.IO;
+
+namespace System.Device.Gpio
+{
+    /// <summary>
+    /// Represents a general-purpose I/O (GPIO) controller.
+    /// </summary>
+    public sealed class GpioController : IDisposable
+    {
+        private Windows.Devices.Gpio.GpioController _controller;
+        private static Windows.Devices.Gpio.GpioPin[] _gpioPins;
+        private static PinEventTypes[] _gpioEvents;
+        private static PinChangeEventHandler[] _gpioPinChange;
+        private static PinEventTypes[] _gpioEventsHappening;
+
+        // this is used as the lock object 
+        // a lock is required because multiple threads can access the GPIO controller
+        static object _syncLock;
+
+        /// <summary>
+        /// Initializes a new instance of the System.Device.Gpio.GpioController class that
+        /// will use the logical pin numbering scheme as default.
+        /// </summary>
+        public GpioController()
+        {
+            GetController();
+        }
+
+        /// <summary>
+        /// Initializes a new instance of the System.Device.Gpio.GpioController class that
+        /// will use the specified numbering scheme. The controller will default to use the
+        /// driver that best applies given the platform the program is executing on.
+        /// </summary>
+        /// <param name="numberingScheme">The numbering scheme used to represent pins provided by the controller.</param>
+        public GpioController(PinNumberingScheme numberingScheme) : this()
+        {
+            NumberingScheme = numberingScheme;
+        }
+
+        private void GetController()
+        {
+            if (_syncLock == null)
+            {
+                _syncLock = new object();
+            }
+
+            lock (_syncLock)
+            {
+                _controller = Windows.Devices.Gpio.GpioController.GetDefault();
+                if (_gpioPins == null)
+                {
+                    _gpioPins = new Windows.Devices.Gpio.GpioPin[_controller.PinCount];
```

Comment:
OK, so I'll use the other implementation based on collections. It's a bit slower because of having to go thru the collection but it works fine as well and definitely less memory intensive. 
In native, we will really have to thing how to make it the best way. Also linked to the way we should manage this on the .NET IoT side. Couple of discussions on evolution of this GPIO API.

Reply:
Understood. Suggest that we move this to full integration with native. It will be more efficient, performance and memory wise. And it will save you some work, for sure! 😉